[tune] document scalability best practices (k8s, scalability thresholds) (#14566)
ray-project/ray: doc/source/tune/api_docs/scalability.rst

.. _tune-scalability:

Scalability and overhead benchmarks
===================================

> We conducted a series of microbenchmarks where we tested the scalability of Ray Tune and tried to reason about
> performance overhead we observed. The results from these benchmarks are reflected in the documentation,

Reviewer: ```suggestion
We conducted a series of microbenchmarks where we evaluated the scalability of Ray Tune and analyzed the
performance overhead we observed. The results from these benchmarks are reflected in the documentation,
```

e.g. when we make suggestions on :ref:`how to remove performance bottlenecks <tune-bottlenecks>`.

This page gives an overview over the experiments we did. For each of these experiments, the goal was to
examine the total runtime of the experiment and address issues when the observed overhead compared to the
minimal theoretical time was too high (e.g. more than 20% overhead).

> 

Reviewer: Maybe you want to note here that you disabled syncing/extra loggers for these experiments?

> .. list-table:: TrialScheduler Feature Compatibility Matrix

Reviewer: change table name?

     - 96
     - 10
     - 1
     - 96
     - 100
     - 168.94
   * - `Network communication overhead <https://github.com/ray-project/ray/blob/master/release/tune_tests/scalability_tests/workloads/test_network_overhead.py>`_
     - 200
     - 1
     - 200
     - 2
     - 300
     - 2280.82
   * - `Long running, 3.75 GB checkpoints <https://github.com/ray-project/ray/blob/master/release/tune_tests/scalability_tests/workloads/test_long_running_large_checkpoints.py>`_
     - 16
     - | Results: 1/60
       | Checkpoint: 1/900
     - 1
     - 16
     - 86,400
     - 88687.41
   * - `XGBoost parameter sweep <https://github.com/ray-project/ray/blob/master/release/tune_tests/scalability_tests/workloads/test_xgboost_sweep.py>`_
     - 16
     - ?
     - 16
     - 64
     - ?
     - 3903
   * - `Durable trainable <https://github.com/ray-project/ray/blob/master/release/tune_tests/scalability_tests/workloads/test_durable_trainable.py>`_
     - 16
     - | 10/60
       | with 10MB CP
     - 16
     - 2
     - 300
     - 392.42

Below we discuss some insights on results where we observed much overhead.

Result throughput
-----------------

> Result throughput is limited by the time it takes to process results. When a trial reports results, it only

Reviewer: Maybe start with `What is result throughput`?

continues training once the trial executor re-triggered the remote training function. If many trials report
results at the same time, each subsequent remote training call is only triggered after handling that trial's
results.

To speed the process up, Ray Tune adaptively buffers results, so that trial training is continued earlier if
many trials are running in parallel and report many results at the same time. Still, processing hundreds of
results per trial for dozens or hundreds of trials can become a bottleneck.

**Main insight**: Ray Tune will throw a warning when trial processing becomes a bottleneck. If you notice
that this becomes a problem, please follow our guidelines outlined :ref:`in the FAQ <tune-bottlenecks>`.
Generally, it is advised to not report too many results at the same time. Consider increasing the report

> intervals.

Reviewer: Consider increasing the report intervals by ...

Below we present more detailed results on the result throughput performance.

Many concurrent trials
""""""""""""""""""""""
In this setup, loggers (CSV, JSON, and TensorboardX) and trial synchronization are disabled, except when

> explicitly noted.

Author: I'll keep the notes here but added a note at the table as well

+-------------+--------------------------+---------+---------------+------------------+---------+
| 256         | 20                       | 16      | 64            | 100s             | 146.20  |
+-------------+--------------------------+---------+---------------+------------------+---------+
| 256         | 10                       | 16      | 64            | 100s             | 113.40  |
+-------------+--------------------------+---------+---------------+------------------+---------+
| 256         | 10 with logging          | 16      | 64            | 100s             | 436.12  |
+-------------+--------------------------+---------+---------------+------------------+---------+
| 256         | 0.1 with logging         | 16      | 64            | 100s             | 106.75  |
+-------------+--------------------------+---------+---------------+------------------+---------+

Many results on a single node
"""""""""""""""""""""""""""""
In this setup, loggers (CSV, JSON, and TensorboardX) are disabled, except when
explicitly noted.

On a single node, throughput seems to be a bit higher. With logging, handling 1000 results per second

> seems acceptable in terms of overhead, though you should probably still target for a lower number.

Reviewer: Can you describe the experiment you ran here? Something like:

```
In this experiment, we measured the scalability of the system under different trial reporting frequencies (changing the number of results per second per trial).
